go-client/client/tablehandle.go

```diff
@@ -0,0 +1,291 @@
+package client
+
+import (
+	"context"
+	"log"
+
+	"github.com/apache/arrow/go/v8/arrow"
+
+	tablepb2 "github.com/deephaven/deephaven-core/go-client/internal/proto/table"
+	ticketpb2 "github.com/deephaven/deephaven-core/go-client/internal/proto/ticket"
+)
+
+// A reference to a table stored on the deephaven server.
+//
+// It should be eventually released using Release once it is no longer needed.
+type TableHandle struct {
+	client   *Client
+	ticket   *ticketpb2.Ticket
+	schema   *arrow.Schema
+	size     int64
+	isStatic bool
+}
+
+func newTableHandle(client *Client, ticket *ticketpb2.Ticket, schema *arrow.Schema, size int64, isStatic bool) *TableHandle {
+	return &TableHandle{
+		client:   client,
+		ticket:   ticket,
+		schema:   schema,
+		size:     size,
+		isStatic: isStatic,
+	}
+}
+
+// Returns true if this table does not change over time.
+// This will be false for things like streaming tables or timetables.
+func (th *TableHandle) IsStatic() bool {
+	return th.isStatic
+}
+
+// Downloads the current state of the table on the server and returns it as a Record.
+//
+// If a Record is returned successfully, it must be freed later with `record.Release()`
+func (th *TableHandle) Snapshot(ctx context.Context) (arrow.Record, error) {
+	return th.client.snapshotRecord(ctx, th.ticket)
+}
+
+// Creates a new query based on this table. Table operations can be performed on query nodes to create new nodes.
+// A list of query nodes can then be passed to client.ExecQuery() to execute the entire query as a single call.
+func (th *TableHandle) Query() QueryNode {
+	qb := newQueryBuilder(th.client, th)
+	return qb.curRootNode()
+}
+
+// Releases this table handle's resources on the server. The TableHandle is no longer usable after Release is called.
+// Ensure that no other methods (including ones in other goroutines) are using this TableHandle.
+func (th *TableHandle) Release(ctx context.Context) error {
+	if th.client != nil {
+		err := th.client.release(ctx, th.ticket)
+		if err != nil {
+			// This is logged because most of the time this method is used with defer,
+			// which will discard the error value.
+			log.Println("unable to release table:", err.Error())
+			return err
+		}
+
+		th.client = nil
+		th.ticket = nil
+		th.schema = nil
+	}
+	return nil
+}
+
+// Returns a new table without the given columns.
+func (th *TableHandle) DropColumns(ctx context.Context, cols ...string) (*TableHandle, error) {
```

Comment:
Is there a good Go design pattern that would allow us to chain the table query operations while still properly handling errors?  Go has been talking about improving error handling for years, but I don't know if they ever adopted any of the proposed changes.

e.g. https://levelup.gitconnected.com/golang-2-0-draft-feature-error-handling-c0a2332b9162

Reply:
Doesn't look like anything has been done on that front. I have looked into "monadic" error handling a little bit, though exactly how it would work is tricky. The other issue with chaining table operations is that the intermediate tables all have to be released individually—and since we can't know what tables are actually referenced, we can't do that automatically.
One possible option is to copy how Java does it, and allow queries to be executed either serially or in batch mode, since the query API already is fairly convenient.

Reply:
Sharp-edged probably makes sense if we print a stack.

Reply:
I've realized, actually: All client tables are released when the client closes. So forgetting about a bunch of table handles and then calling `client.Close()` is a perfectly valid way to clean up resources—unless we implement sharp-edged finalizers.

Perhaps we could have some kind of optional client setting like `ForceTableCleanup`, so that users that know they're creating a long-running task can get errors, and users with short-lived sessions don't have to worry. I'm not sure this is a feature worth adding to our initial release, though (or if that's even the right way to go about it).